_tools/types/main.go

```diff
@@ -0,0 +1,341 @@
+		wg sync.WaitGroup
+	)
+	for i := range drivers {
+		wg.Add(1)
+		go func(driver *driverStats) {
+			defer wg.Done()
+			maybeCloneOrPullThrottled(driver, pf)
+		}(drivers[i])
+	}
+	wg.Wait()
+	return nil
+}
+
+// maybeCloneOrPullThrottled is concurent-friendly verison of maybeCloneOrPull.
+// It "throttles" git operarions, based on given chanel capacity.
+func maybeCloneOrPullThrottled(d *driverStats, concurent chan int) error {
+	concurent <- 1
+	defer func() {
+		<-concurent
+	}()
+
+	return maybeCloneOrPull(d)
+}
+
+func maybeCloneOrPull(d *driverStats) error {
+	repoPath := path.Join(reposRootPath, d.path)
+	_, err := os.Stat(repoPath)
+	if os.IsNotExist(err) {
+		fmt.Fprintf(os.Stderr, "%s does not exist, cloning from %s\n", repoPath, d.url)
+		cmd := exec.Command("git", "clone", d.url+".git")
+		cmd.Dir = reposRootPath
+		err = cmd.Run()
+		if err != nil {
+			return err
+		}
+		return nil
+	} else if err != nil {
+		fmt.Fprintf(os.Stderr, "error: %s\n", err)
+		return err
+	}
+
+	fmt.Fprintf(os.Stderr, "%s dir exists, will git pull instead\n", repoPath)
+	cmd := exec.Command("git", "pull", "origin", "master")
+	cmd.Dir = repoPath
+	err = cmd.Run()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type uastType struct {
+	name string
+}
+
+// isUsedIn checks if current UAST type is used in the given package.
+func (u *uastType) isUsedIn() {
+	// TODO
+}
+
+// findUASTTypesInSDK finds all types from SDK.
+func findUASTTypesInSDK() []uastType {
+	var out []uastType
+	// TODO: either parse code of the package, iterate all structs
+	// or expose all registered types though SDK and use it here
+	types := []interface{}{
+		uast.Position{},
+		uast.Positions{},
+		uast.Identifier{},
+		uast.String{},
+		uast.Bool{},
+		uast.QualifiedIdentifier{},
+		uast.Comment{},
+		uast.Group{},
+		uast.FunctionGroup{},
+		uast.Block{},
+		uast.Alias{},
+		uast.Import{},
+		uast.RuntimeImport{},
+		uast.RuntimeReImport{},
+		uast.InlineImport{},
+		uast.Argument{},
+		uast.FunctionType{},
+		uast.Function{},
+	}
+	for _, typee := range types {
+		out = append(out, uastType{reflect.TypeOf(typee).String()})
+	}
+	fmt.Fprintf(os.Stderr, "%d uast:* types found in SDK\n", len(out))
+	return out
+}
+
+// analyzeFixtures goes though all fixtures, assuming the driver is cloned.
+// It updates given driverStats with results.
+func analyzeFixtures(driver *driverStats) error {
+	fixDir := path.Join(reposRootPath, driver.path, fixtureDir)
+	fixtureFiles, err := lsDir(fixDir)
+	if os.IsNotExist(err) {
+		driver.skip = true
+		return nil
+	} else if err != nil {
+		return err
+	}
+	driver.fixturesNum += len(fixtureFiles)
+
+	for _, file := range fixtureFiles {
+		if !strings.HasSuffix(file.Name(), fixtureExt) {
+			continue
+		}
+
+		fFile, err := os.Open(path.Join(fixDir, file.Name()))
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "failed to read %s, skipping\n", path.Join(fixDir, file.Name()))
+		}
+
+		scanner := bufio.NewScanner(fFile)
```

Comment:
It looks to me like we have to read the whole file anyway, and these files are relatively small (just human-written source). On that basis, I would advocate a simpler strategy:

```go
data, err := ioutil.ReadFile(filepath.Join(fixDir, file.Name())
if err != nil {
  fmt.Fprintf(os.Stderr, "Unable to read %q, skipping\n", file.Name())
  continue
}
for _, line := range strings.Split(string(data), "\n") {
   …
}
```

It does cause the whole file to be read in, but for small files it may actually do _fewer_ allocations (not that this is perf critical anyway, since it's an occasional generation step). And it's simpler, not having to worry about whether the filehandle has been closed in various error conditions.

Reply:
Those fixtures may be relatively large, actually (tens of MB). So it should be faster to scan.

Reply:
Was considering this, but decided in favor of buffered io, as I did not measure the performance of either approache.

Reply:
Good point about the size. That said: I doubt it would be any faster—we're reading the whole file either way—the main savings is RAM. I don't think we really care about RAM usage here, it's a command-line tool run once in a while to update the stats, and we're only processing one by one.

Reply:
Was changed to ioutil in 557d9b6

Reply:
Thank you for sharing the assumptions and performance expectations though! It's such "layered" subject in modern systems that developing an intuition really takes a collective effort!

Did a few tests, on the current usage patten (scan by line + check & match)
  - `ioutil` version is consistent outperforming the `bufio` one ±2.20 sec
 - `bufio` version is ±2.40 sec

these are by no means an accurate benchmarks, just an observation on the time of processing all files, avg on 5 runs.

Reply:
Results of removing variance due to network driver discovery and properly benchmarking `analyzeFixtures` on all 10 drivers:

```
Benchmark_bufio_ScanMatch-4   	   30000	     45404 ns/op	    1888 B/op	      50 allocs/op
Benchmark_ioutil_ScanMatch-4   	   30000	     41721 ns/op	    1888 B/op	      50 allocs/op
```

 - allocations are exactly the same
 - ioutil is a bit faster

🤔

Reply:
I don't know for sure, but I suspect the issue could be if your benchmark is using a file too small to really trigger any differences.

Reply:
>  issue could be if your benchmark is using a file too small

Hm.. so far it uses the same data as for report generation: iterates over all 1001 `./fixtures/*.sem.uast` files in 10 different drivers. I'll dig deeper tomorrow and report back!

Reply:
The bytes-per-op are suspiciously small, and the allocations suspiciously uniform. I don't know what's going on there, but I think there must be something wrong in the way the benchmark is getting its data. (This is by no means crucial, so apart from curiosity please don't let this hold you up 😄).

Reply:
Sorry for delay, here is an updated benchmark on single 500Mb UAST

```
Benchmark_bufio_ScanMatch_one-4		10	2390952772 ns/op	842338523 B/op	 9027097 allocs/op
Benchmark_ioutil_ScanMatch_one-4	10	1417128306 ns/op	1432769024 B/op	 4214468 allocs/op
```

`bufio` uses less memory but allocates more. `ioutil` is almost x2 faster but uses more memory.